### PR TITLE
Add Kyber post-quantum key encapsulation mechanism

### PR DESCRIPTION
**Pull request by bwesterb**
Implementation of the IND-CCA2 post-quantum secure key encapsulation mechanism (KEM) CRYSTALS-Kyber, as submitted to the third round of the NIST Post-Quantum Cryptography (v3.02/"draft00"), and selected for standardisation.

Kyber will likely change before final standardisation.

This implementation is based on the one I wrote [in Go](https://github.com/cloudflare/circl/tree/main/kem/kyber).

### CODE REVIEW

**Comment by bwesterb** [comment]
@jedisct1

**Comment by jedisct1** [comment]
Let's discuss this in your Zig fork :)

**Comment by bwesterb** [comment]
On M2 Pro. Essentially the reference implementation:

```
Did 12000 Kyber512 encap operations in 1034796us (11596.5 ops/sec)
Did 10344 Kyber512 decap operations in 1062489us (9735.6 ops/sec)
```

This implementation:

```
         kyber512:      98172 encaps/s
         kyber512:      92426 decaps/s
```

I think the difference is caused by Zig's Keccak's permutation f.

For comparison with classical crypto, in Zig:

```
           x25519:      33924 exchanges/s
```

and BoringSSL:

```
Did 4598 Curve25519 arbitrary point multiplication operations in 1052611us (4368.2 ops/sec)
```

**Comment by bwesterb** [comment]
> ```
>          kyber512:      98172 encaps/s
>          kyber512:      92426 decaps/s
> ```

Wait... those are 10x higher. Let me double check.

**Comment by jedisct1** [comment]
This is rad!

And there's still plenty of room for optimization on the Keccak side.

**Review by jedisct1 — APPROVED** [review]

File: lib/std/crypto/kyber_d00.zig [hunk 1]
```diff
@@ -0,0 +1,1780 @@
...
+        }
+
+        return p;
+    }
+
+    // Sample p from a centered binomial distribution with n=2η and p=½ - viz:
+    // coefficients are in {-η, …, η} with probabilities
+    //
+    //  {ncr(0, 2η)/2^2η, ncr(1, 2η)/2^2η, …, ncr(2η,2η)/2^2η}
+    fn noise(comptime eta: u8, nonce: u8, seed: *const [32]u8) Poly {
```
Comment (bwesterb) [inline lib/std/crypto/kyber_d00.zig:1156]: Peter, here :)

**Comment by jedisct1** [comment]
CI failure is unrelated; it's due to `d6e48ab`.

**Comment by andrewrk** [comment]
> CI failure is unrelated; it's due to [d6e48ab](https://github.com/ziglang/zig/commit/d6e48abde87400a8a4851c7ab8c918005d81d058)

It's a nondeterminism issue; probably in the macOS linker code. It's not the fault of d6e48abde87400a8a4851c7ab8c918005d81d058.

**Pull request merged**